soroban-env-host/src/host.rs

```diff
@@ -1684,18 +1698,92 @@ impl VmCallerEnv for Host {
         _vmcaller: &mut VmCaller<Host>,
         k: RawVal,
         v: RawVal,
+        t: StorageType,
+        f: RawVal,
     ) -> Result<Void, HostError> {
-        let key = self.contract_data_key_from_rawval(k)?;
-        let data = LedgerEntryData::ContractData(ContractDataEntry {
-            contract_id: self.get_current_contract_id_internal()?,
-            key: self.from_host_val(k)?,
-            val: self.from_host_val(v)?,
-        });
-        self.0.storage.borrow_mut().put(
-            &key,
-            &Host::ledger_entry_from_data(data),
-            self.as_budget(),
-        )?;
+        let flags = if f.is_void() {
+            0
+        } else {
+            self.u32_from_rawval_input("f", f)?
+        };
+
+        if ((flags as u64) & !MASK_CONTRACT_DATA_FLAGS_V20) != 0 {
+            return Err(self.err(
+                ScErrorType::Value,
+                ScErrorCode::InvalidInput,
+                "invalid flags",
+                &[],
+            ));
+        }
+
+        let storage_type: ContractDataType = t.try_into()?;
+        let key = self.contract_data_key_from_rawval(k, storage_type)?;
+        if self.0.storage.borrow_mut().has(&key, self.as_budget())? {
+            let mut current = (*self.0.storage.borrow_mut().get(&key, self.as_budget())?)
+                .metered_clone(&self.0.budget)?;
+
+            match current.data {
+                LedgerEntryData::ContractData(ref mut entry) => match entry.body {
+                    ContractDataEntryBody::DataEntry(ref mut data) => {
+                        data.val = self.from_host_val(v)?;
+                        if flags == 0 {
+                            data.flags = 0;
+                        } else {
+                            data.flags |= flags;
```

Comment:
It seems a little weird to me that a `put` of flags does not just overwrite the old flags (as it does the value), and a little weirder still that it has two _separate_ meanings (reset if zero, bitwise-or if nonzero). Is there a compelling use-case for these semantics? I would not guess them, though I see they're what got documented in env.json, they seem surprising to me.

Reply:
The bitwise-or is pretty confusing, so I'll change this to just overwrite instead. I think the optional semantics makes sense though so you only write the flags field when you intend to.